python/lsst/dax/distribution/DataIngest.py

```diff
@@ -0,0 +1,297 @@
+        id = r_json['databases'][db_name]['transactions'][0]['id']
+        print("transaction id=", id, " r_json=", r_json)
+        return True, id
+
+    def endTransaction(self, transaction_id, abort):
+        """ End the ingest transaction
+
+        Parameters
+        ----------
+        transaction_id : int
+            The transaction id number.
+        abort : bool
+            True if the transaction should be rolled back.
+
+        Return
+        ------
+        success : bool
+            True if the operation was successful.
+        status : int
+            Status code from the put operation.
+        r_json : json or None
+            json data from the put operation if status was 200.
+        """
+        cmd = 'ingest/trans/' + str(transaction_id) + '?abort='
+        if abort:
+            cmd += '1'
+        else:
+            cmd += '0'
+        success, status, r_json = self._putToIngest(cmd, {'auth_key':self._auth_key})
+        if not success:
+            print("ERROR ending transaction id=", transaction_id, "abort=", abort, "status=", status,
+                  "r_json=", r_json)
+        return success, status, r_json
+
+    def getChunkTargetAddr(self, transaction_id, chunk_id):
+        """ Get the host and port number of the ingest worker for this chunk.
+
+        Parameters
+        ----------
+        transaction_id : int
+            Ingest transaction id number.
+        chunk_id : int
+            Chunk id number.
+
+        Return
+        ------
+        host : str
+            Host name of the ingest worker.
+        port : int
+            Port number of the ingest worker.
+        """
+        cmd = 'ingest/chunk'
+        jdata = {"transaction_id":transaction_id,"chunk":chunk_id,"auth_key":self._auth_key}
+        success, r_json = self._postToIngest(cmd, jdata)
+        if not success:
+            print("ERROR failed to get chunk target address", jdata, " r_json=", r_json)
+            raise RuntimeError('Transaction ' + transaction_id
+                               + ' failed to get target address for chunk ' + chunk_id)
+        host = r_json['location']['host']
+        port = r_json['location']['port']
+        return host, port
+
+    def sendChunkToTarget(self, host, port, transaction_id, table, f_path):
+        """ Send the file to the ingest worker.
+
+        Parameters
+        ----------
+        host : str
+            Host name for the ingest worker.
+        port : int
+            Port number for the ingest worker.
+        transation_id : int
+            Ingest transaction id number.
+        table : str
+            The name of the table that f_path data should be added to.
+        f_path : str
+            Full path to the file to ingest.
+
+        Note
+        ----
+        This calls the external program 'qserv-replica-file-ingest' to
+        actually send the file.
+        """
+        cmd = ('qserv-replica-file-ingest FILE ' + host + ' ' + str(port) + ' '
```

Comment:
For the record, this operation may be less efficient than sending many files in batch mode as explained here: https://confluence.lsstcorp.org/pages/viewpage.action?pageId=133333850#UserguidefortheQservIngestsystem(APIversion1)-Uploadingchunkcontributions and here: https://confluence.lsstcorp.org/display/DM/Ingest%3A+11.2.2.+qserv-replica-file-ingest
Though, this may require to improve your workflow

Reply:
I did it this because it made the rest of the code much simpler and it will take a substantial code change to alter it. I guess a reasonable question is how much more efficient? I expect files to be many megabytes or gigabytes in most cases and I hope the overhead to ingest a file is very small in comparison. The need for this is getting to be urgent, so I'd like to see if optimization is needed before spending time on it.

Reply:
It all depends on the total number of files and file sizes, and also how you're going to submit your contributions. If you do everything from inside a container then an overhead for launching the `qsev-replica-file-ingest` binary may be small. However if you're going to launch a Docker container for a single file ingest then it would be a very different story. 
Your choices (which of those two options to use: `FILE` or `FILE-LIST`) would also depend on an organization of your workflow. From my reading of your comments in the JIRA ticket on tis code, it looks like your workflow is around each chunk:
# you have a list of chunks to be generated
# you have multiple "clients" each asking your server the next available chunk to process
# and each clients is working on one chunk at a time, generates the "fake" data, partitions the chunk file to make overlaps and ingests the resulting products into Qserv
In this model `FILE` looks like a natural choice.  The main benefit of your approach is that it won't require any significant temporary storage to accumulate a lot of data during steps 1 and 2 above.
An alternative approach would be to let each worker generate N chunks, partition them into some output folder and then initiate the ingest phase using `FILE-LIST`. This approach will require you to be cautious regarding the amount of free space available in your temporary storage area.
In the end it's your call. You will see how it all works when you will be doing large-scale generation/ingest. You may always optimize your workflow later.
Just one last word of caution: please, avoid starting too many "super-transactions" in your workflow. It would be reasonable to keep the total number of transactions you have during the whole ingest campaign under 100 or so. See some notes on this subject at: https://confluence.lsstcorp.org/pages/viewpage.action?pageId=133333850#UserguidefortheQservIngestsystem(APIversion1)-Multipletransactions